File: app/src/main/java/za/co/woolworths/financial/services/android/ui/activities/voc/VoiceOfCustomerActivity.java
package za.co.woolworths.financial.services.android.ui.activities.voc;

import android.content.Intent;
import android.os.Bundle;
import android.view.MenuItem;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.appcompat.app.ActionBar;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.fragment.app.Fragment;
import androidx.navigation.NavController;
import androidx.navigation.NavGraph;
import androidx.navigation.fragment.NavHostFragment;

import com.awfs.coordination.R;

import org.jetbrains.annotations.NotNull;

import java.util.List;

import za.co.woolworths.financial.services.android.util.Utils;

public class VoiceOfCustomerActivity extends AppCompatActivity implements VoiceOfCustomerInterface {

    private NavController navigationHost = null;
    private Toolbar mPrefsToolbar;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        // TODO: add param to be either right to left, or bottom to top
        overridePendingTransition(R.anim.slide_from_right, R.anim.slide_to_left);

        Utils.updateStatusBarBackground(this);
        setContentView(R.layout.activity_voice_of_customer);

        NavHostFragment vocNavHostFrag = (NavHostFragment) getSupportFragmentManager().findFragmentById(R.id.vocNavHostFrag);
        if (vocNavHostFrag != null) {
            navigationHost = vocNavHostFrag.getNavController();
        }
        mPrefsToolbar = findViewById(R.id.mVocToolbar);
        setActionBar();
        setNavHostStartDestination();
    }

    private void setActionBar() {
        setSupportActionBar(mPrefsToolbar);
        ActionBar mActionBar = getSupportActionBar();
        if (mActionBar != null) {
            mActionBar.setDisplayHomeAsUpEnabled(true);
            mActionBar.setDisplayShowTitleEnabled(false);
            mActionBar.setDisplayUseLogoEnabled(false);
            mActionBar.setHomeAsUpIndicator(R.drawable.back24);
        }
    }

    private void setNavHostStartDestination() {
        if (navigationHost == null) {
            return;
        }
        NavGraph graph = navigationHost.getGraph();
        graph.setStartDestination(R.id.surveyVocFragment);

        navigationHost.setGraph(graph, getIntent() != null ? getIntent().getExtras() : null);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);

        NavHostFragment navHostFragment = (NavHostFragment) getSupportFragmentManager().getFragments().get(0);
        if (navHostFragment != null) {
            List<Fragment> childFragments = navHostFragment.getChildFragmentManager().getFragments();
            for (Fragment fragment : childFragments) {
                fragment.onActivityResult(requestCode, resultCode, data);
            }
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                onBackPressed();
                return true;
        }
        return false;
    }

    @Override
    public void onBackPressed() {
        if (navigationHost.getCurrentDestination() == null) {
            return;
        }

        switch (navigationHost.getCurrentDestination().getId()) {
            case R.id.surveyVocFragment:
                finishActivity();
                break;
            default:
                navigationHost.popBackStack();
                break;
        }
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        // permission was granted, yay! Do the
        // contacts-related task you need to do.
        getCurrentFragment().onRequestPermissionsResult(requestCode, permissions, grantResults);
    }

    private Fragment getCurrentFragment() {
        Fragment navHostFragment = getSupportFragmentManager().findFragmentById(R.id.vocNavHostFrag);
        return navHostFragment == null ? null : navHostFragment.getChildFragmentManager().getFragments().get(0);
    }

    private void finishActivity() {
        finish();
        overridePendingTransition(R.anim.slide_from_left, R.anim.slide_to_right);
    }

    @Override
    public void setToolbarTitle(@NotNull String titleTxt) {
        if (mPrefsToolbar == null) {
            return;
        }
        TextView title = mPrefsToolbar.findViewById(R.id.toolbarText);
        title.setText(titleTxt);
    }

    @Override
    public void setToolbarTitleGravity(int gravity) {
        if (mPrefsToolbar == null) {
            return;
        }
        TextView title = mPrefsToolbar.findViewById(R.id.toolbarText);
        title.setGravity(gravity);
    }
}

File: app/src/main/java/za/co/woolworths/financial/services/android/ui/activities/voc/VoiceOfCustomerActivity.kt
package za.co.woolworths.financial.services.android.ui.activities.voc

import android.os.Bundle
import android.view.MenuItem
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.navigation.NavController
import androidx.navigation.fragment.NavHostFragment
import com.awfs.coordination.R
import com.awfs.coordination.databinding.ActivityVoiceOfCustomerBinding
import za.co.woolworths.financial.services.android.contracts.FirebaseManagerAnalyticsProperties
import za.co.woolworths.financial.services.android.ui.views.actionsheet.GenericActionOrCancelDialogFragment
import za.co.woolworths.financial.services.android.util.KeyboardUtils
import za.co.woolworths.financial.services.android.util.Utils

class VoiceOfCustomerActivity : AppCompatActivity(R.layout.activity_voice_of_customer), VoiceOfCustomerInterface, GenericActionOrCancelDialogFragment.IActionOrCancel {

    companion object {
        const val EXTRA_SURVEY_DETAILS = "extraSurveyDetails"
        const val EXTRA_SURVEY_ANSWERS = "extraSurveyAnswers"
        const val DIALOG_SKIP_ID = 1
        const val DEFAULT_VALUE_RATE_SLIDER_MIN = 1
        const val DEFAULT_VALUE_RATE_SLIDER_MAX = 11
    }

    private lateinit var binding: ActivityVoiceOfCustomerBinding
    private var navigationHost: NavController? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        overridePendingTransition(R.anim.slide_up_anim, R.anim.stay)

        Utils.updateStatusBarBackground(this)
        binding = ActivityVoiceOfCustomerBinding.inflate(layoutInflater)
        setContentView(binding.root)

        val vocNavHostFrag = supportFragmentManager.findFragmentById(R.id.vocNavHostFrag) as NavHostFragment?
        if (vocNavHostFrag != null) {
            navigationHost = vocNavHostFrag.navController
        }

        setActionBar()
        setNavHostStartDestination()

        // Hide keyboard in case it was visible from a previous screen
        KeyboardUtils.hideKeyboardIfVisible(this)
    }

    private fun setActionBar() {
        setSupportActionBar(binding.vocToolbar)
        supportActionBar?.apply {
            setDisplayHomeAsUpEnabled(true)
            setDisplayShowTitleEnabled(false)
            setDisplayUseLogoEnabled(false)
            setHomeAsUpIndicator(R.drawable.back24)
        }
    }

    private fun setNavHostStartDestination() {
        val graph = navigationHost?.graph ?: return
        graph.startDestination = R.id.surveyVocFragment
        navigationHost?.setGraph(graph, if (intent != null) intent.extras else null)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                onBackPressed()
                return true
            }
        }
        return false
    }

    override fun onBackPressed() {
        Utils.triggerFireBaseEvents(FirebaseManagerAnalyticsProperties.VOC_SKIP, this)
        val currentDestination = navigationHost?.currentDestination ?: return
        when (currentDestination.id) {
            R.id.surveyVocFragment -> finishActivity()
        }
    }

    fun finishActivity() {
        finish()
        overridePendingTransition(R.anim.stay, R.anim.slide_down_anim)
    }

    override fun setToolbarSkipVisibility(show: Boolean) {
        binding.apply {
            if (vocToolbar == null) return
            tvSkipSurvey.visibility = if (show) View.VISIBLE else View.GONE
            tvSkipSurvey.setOnClickListener { onSkipSurveyClicked() }
        }
    }

    fun hideToolbarBackButton() {
        supportActionBar?.setDisplayHomeAsUpEnabled(false)
    }

    private fun onSkipSurveyClicked() {
        val dialog = GenericActionOrCancelDialogFragment.newInstance(
                dialogId = DIALOG_SKIP_ID,
                title = getString(R.string.voc_skip_dialog_title),
                desc = getString(R.string.voc_skip_dialog_desc),
                actionButtonText = getString(R.string.voc_skip_dialog_action),
                cancelButtonText = getString(R.string.voc_skip_dialog_cancel),
                this
        )
        dialog.show(supportFragmentManager, GenericActionOrCancelDialogFragment::class.java.simpleName)
    }

    override fun onDialogActionClicked(dialogId: Int) {
        if (dialogId == DIALOG_SKIP_ID) {
            Utils.triggerFireBaseEvents(FirebaseManagerAnalyticsProperties.VOC_SKIP, this)
            finishActivity()
        }
    }
}